// file-script-handler/lib/fileScriptHandler/imageSizeHandler.js
/** @module file-script-handler/imageSizeHandler */
'use strict';

const fs = require('fs');
const path = require("path");
const im = require('imagemagick');
const Q = require('q');
const DachshundError = require('dachshund-error');
const resourceTypes = require('resource-types');
const mime = require('mime');
const temp = require('temp').track();
const sha256 = require('sha256');
const repositoryManager = require('repository-manager');
const stringUtils = require('string-utils');
const imageSizeHandler = exports;
const nodeStatic = require('node-static');
const fileServer = new nodeStatic.Server({ cache: false });
const assert = require('assert');

imageSizeHandler.test = function(pathInfo) {
	let extension = pathInfo.extension.toLowerCase();
	let correctExtension = extension == "jpg" || extension == "jpeg" || extension == "png" || extension == "gif";
	let selectorIsDefined = stringUtils.isNotEmpty(pathInfo.selector);
	return selectorIsDefined && correctExtension;
};

/**
 * Resolves the image thumbnail or creates it, for the case that it doesn't exist
 * @param req
 * @param resp
 * @param pathInfo
 */
imageSizeHandler.handle = function(req, resp, pathInfo) {
	try
	{
		let thumbProperties = new ImageProperties(pathInfo);
	} catch (err)
	{
		resp.writeHead(500, {"content-type": "text/plain; charset=utf-8"});
		resp.write(err.toString());
		resp.end();
		return;
	}
	req.session.resolve(thumbProperties.sourceFilePath).then(function(imageParentResource) {
		let deferred = Q.defer();

		//Step 1: Check if the resource exists and access is allowed
		if (!resourceTypes.isResource(imageParentResource))
		{
			deferred.reject(new DachshundError("404 Resource not found", 404));
		}
		else
		{
			//Step 2: Check if source image exists
			imageParentResource.fileExists(thumbProperties.sourceFileName).then(function(sourceFileExists) {
				if (!sourceFileExists)
				{
					deferred.reject(new DachshundError("404 Resource not found", 404));
				}
				else
				{
					deferred.resolve(imageParentResource);
				}
			});
		}
		return deferred.promise;
	}).then(function(imageParentResource) {
		let deferred = Q.defer();

		//Step 3: Check if the thumb image is already created
		imageParentResource.fileExists(thumbProperties.thumbFileName).then(function(thumbFileExists) {
			if (thumbFileExists)
			{
				//Step 4a: Serve the image when it exists
				deferred.resolve(thumbProperties);
			}
			else
			{
				//Step 4b: Create the thumbnail and serve it when it's created
				createThumbnail(thumbProperties).then(function() {
					deferred.resolve(thumbProperties);
				}).fail(deferred.reject);
			}
		});

		return deferred.promise;
	}).then(function(thumbProperties) {
		let thumbImagePath = thumbProperties.sourceFilePath + "/" + thumbProperties.thumbFileName;
		req.session.resolve(thumbImagePath).then(function(thumbFileResource) {
			let thumbImageFilePath = "." + (thumbFileResource.source || thumbImagePath);
			let mimetype = mime.lookup(pathInfo.extension);
			fileServer.serveFile(thumbImageFilePath, 200, {"content-type": mimetype}, req, resp);
		});
	}).fail(function(err) {
		resp.writeHead(err.id || 500, {"Content-Type": "text/plain; charset=utf-8"});
		resp.write(err.toString());
		resp.end();
	});
};

/**
 * Creates the thumbnail and stores it at the parent resource where the original image belongs to
 * @param {ImageProperties} thumbProperties Information about the thumbnail process
 * @returns {*}
 */
const createThumbnail = function(thumbProperties) {
	let deferred = Q.defer();

	let completeSourcePath = thumbProperties.sourceFilePath + "/" + thumbProperties.sourceFileName;
	let sourceFilePath = "";

	//Step 1: Access the source image
	repositoryManager.resolve(completeSourcePath).then(function(sourceImageResource) {
		sourceFilePath = sourceImageResource.source || completeSourcePath;

		//Step 2: Read the the image exif information
		return Q.nfcall(im.readMetadata, process.cwd() + sourceFilePath);
	}).then(function(metadata) {
		let deferred = Q.defer();

		//Step 3a: Set the size depending on the orientation of the image
		let size = thumbProperties.getSizeVertical();
		if (metadata.exif != undefined && metadata.exif.orientation != undefined)
		{
			// All orientations great as 4 are a pictures is
			// horizontal instead of vertical
			// Details are here documented:
			// http://sylvana.net/jpegcrop/exif_orientation.html
			if (metadata.exif.orientation > 4)
			{
				size = thumbProperties.getSizeHorizontal();
			}
		}

		//Step 3b: If the minimum size instead of the maximum is requested, the requested size has to be prefixed for image magic
		if (thumbProperties.type == "min")
		{
			size += "^";
		}

		//Step 4: Create a temporary file where the image is written to, until it will be added to the resource
		Q.nfcall(temp.open, sha256(completeSourcePath)).then(function(tempInfo) {
			//Step 5: Create the image thumbnail and write it into the temporary file
			Q.nfcall(im.convert, [ process.cwd() + sourceFilePath, '-resize', size, tempInfo.path]).then(function() {
				deferred.resolve(tempInfo.path);
			}).fail(deferred.reject);
		}).fail(deferred.reject);

		return deferred.promise;
	}).then(function(thumbnailPath) {
		let deferred = Q.defer();
		//Step 6: Switch the orientation of the image, if it was taken vertical
		Q.nfcall(im.convert, [ '-auto-orient', thumbnailPath, thumbnailPath]).then(function() {
			deferred.resolve(thumbnailPath);
		}).fail(deferred.reject);
		return deferred.promise;
	}).then(function(thumbnailPath) {
		let deferred = Q.defer();
		//Step 7: Resolve the resource where the image will be attched to
		repositoryManager.resolve(thumbProperties.sourceFilePath).then(function(imageParentResource) {
			//Step 8: Attach thumbnail to the parent resource of the source image
			let thumbnailReadStream = fs.createReadStream(thumbnailPath);
			imageParentResource.addFile(thumbnailReadStream, thumbProperties.thumbFileName).then(deferred.resolve).fail(deferred.reject);
		}).fail(deferred.reject);
		return deferred.promise;
	}).then(deferred.resolve).fail(deferred.reject);

	return deferred.promise;
};

/**
 * Resolves the necessary parameters to create the image thumbnail
 *
 * @param pathInfo RequestPathInfo object resolved from the http request
 */
const ImageProperties = function(pathInfo) {
	assert(stringUtils.isNotEmpty(pathInfo.selector), "Selector isn't defined");
	let selectorParts = pathInfo.selector.split(".");
	if (selectorParts.length >= 2)
	{
		if (selectorParts[selectorParts.length - 2] == "min" || selectorParts[selectorParts.length - 2] == "max")
		{
			this.type = selectorParts[selectorParts.length - 2];
			selectorParts.splice(selectorParts.length - 2, 1);
		}
		this.size = selectorParts[selectorParts.length - 1];
		selectorParts.splice(selectorParts.length - 1, 1);
	}
	else
	{
		this.size = selectorParts[0];
		selectorParts = [];
	}
	assert(stringUtils.isNotEmpty(this.size), "Size isn't defined within request selector.");
	let sizeParts = this.size.split("x");
	assert.equal(sizeParts.length, 2, "Size selector is malformed.");

	this.width = parseInt(sizeParts[0]);
	assert(!isNaN(this.width), "Defined width isn't a number");

	this.height = parseInt(sizeParts[1]);
	assert(!isNaN(this.height), "Defined height isn't a number");

	let completeSourcePath = pathInfo.componentPath + "." + pathInfo.extension;
	if (selectorParts.length > 0)
	{
		completeSourcePath = pathInfo.componentPath + "." + selectorParts.join(".") + "." + pathInfo.extension;
	}
	this.sourceFilePath = completeSourcePath.lastSubstringBefore("/");
	this.sourceFileName = completeSourcePath.lastSubstringAfter("/");
	this.thumbFileName = "." + this.sourceFileName.lastSubstringBefore(".") + "." + this.type + "." + this.size + "." + pathInfo.extension
};

ImageProperties.prototype.type = "max";
ImageProperties.prototype.size = "";
ImageProperties.prototype.height = 0;
ImageProperties.prototype.width = 0;
ImageProperties.prototype.sourceFilePath = "";
ImageProperties.prototype.sourceFileName = "";
ImageProperties.prototype.thumbFileName = "";

ImageProperties.prototype.getSizeVertical = function() {
	return this.width + "x" + this.height;
};

ImageProperties.prototype.getSizeHorizontal = function() {
	return this.height + "x" + this.width;
};
