// repository-session/lib/repositorySession.js
'use strict';

const Q = require('q');
const repositoryManager = require('repository-manager');
const secureResourceManager = require('./secureResourceManager.js');

/**
 * @example
 * prepare.handle = function(req, res, pathInfo, resource, component) {
    const deferred = Q.defer();

	resolveHomepage(resource).then(function(){
		return resolveSubHeadings(req.session, resource);
	}).then(deferred.resolve).fail(deferred.reject);

    return deferred.promise;
};
 *
 * @extends RepositoryManager
 *
 */
class RepositorySession {

	/**
	 * Creates a repository session based on the provided user session. Requests executed with the object will incorporate the access rights of the referenced {@link UserSession}.
	 * @param {UserSession} userSession - The user session which is applied to evaluate if the user has access rights.
	 */
	constructor(userSession) {
		this.user = userSession;
	}

	/**
	 * @inheritDoc
	 */
	createEmptyResource(pathInfo) {
		return repositoryManager.createEmptyResource(pathInfo);
	};

	/**
	 * @inheritDoc
	 */
	createResource(newResourceBasePath, newResourceName) {
		return repositoryManager.createResource(newResourceBasePath, newResourceName);
	};

	/**
	 * @inheritDoc
	 */
	resolve(pathInfo) {
		let deferred = Q.defer();
		let sessionUser = this.user;
		repositoryManager.resolve(pathInfo).then(function(resource) {
			secureResourceManager.secureResource(resource, sessionUser).then(function(resource) {
				deferred.resolve(resource);
			}).fail(deferred.reject);
		}).fail(deferred.reject);
		return deferred.promise;
	};
}


module.exports = RepositorySession;
